File: week-2/find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.js
/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function(s, p) {
    const res = [];
    if (!s || s.length === 0 || p.length > s.length) return res;
    
    const array = new Array(26).fill(0);
    for (let i of p) {
        ++array[i.charCodeAt() - 'a'.charCodeAt()];
    }
    
    let left = 0;
    let right = 0;
    let count = p.length;
    
    while (right < s.length) {
        if (array[s[right].charCodeAt() - 'a'.charCodeAt()] >= 1) {
            --count;
        }
        
        --array[s[right].charCodeAt() - 'a'.charCodeAt()];
        ++right;
        
        if (count === 0) res.push(left);
        
        if (right-left === p.length) {
            if (array[s[left].charCodeAt() - 'a'.charCodeAt()] >= 0) {
                ++count;
            }
            
            ++array[s[left].charCodeAt() - 'a'.charCodeAt()];
            ++left;
        }
    }
    return res;
};

File: week-2/find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if not s or len(s) == 0 or len(p) > len(s): return []
        array = [0]*26
        for i in p:
            array[ord(i) - ord('a')] += 1
            
        left = 0
        right = 0
        count = len(p)
        res = []
        
        while right < len(s):
            if array[ord(s[right]) - ord('a')] >= 1:
                count -= 1
            
            array[ord(s[right]) - ord('a')] -= 1
            right += 1
            
            if count == 0: res.append(left)
                
            if right - left == len(p):
                if array[ord(s[left]) - ord('a')] >= 0:
                    count += 1
                array[ord(s[left]) - ord('a')] += 1
                left += 1
                
        return res
